### ISSUE TITLE
Save result of pseudo-experiments/toys

### ISSUE THREAD

**Issue by marinang** [scikit-hep/hepstats#18]
In #14 the frequentist calculator is introduced which uses toys to build test statistic distributions to compute pvalues for instance.

As the generation of toys + fitting + scanning is quite CPU intensive and takes time, an improvement would be to store the result of the pseudo experiments for instance in an hdf5 file. 
This is currently an issue for testing which have time limits. For each pseudo experiment what need to be stored is:

* the values of POI at generation
* the best fitted values of the POI
* the likelihood evaluated at the best fit POI
* the likelihood evaluated at the POI points you want to scan (for upper limits or confidence intervals)

The results stored can be reused in `hepstats.hypotests` with the frequentist calculator without regenerating pseudo-experiments. This would let the possibility to the users to generate pseudo-experiments outside of `heptstats` as well.

A first design I have in mind is to create a class called `ToyManager` for instance which collect and save the results of the pseudo-experiments and can reopen them.

What do you think @eduardo-rodrigues @mayou36 @HDembinski ?

**Comment by eduardo-rodrigues** [comment]
Hi @marinang, Brian had the same usecase when working in the `scikit-hep` package visual module, and he used fixtures to save material, see `https://github.com/scikit-hep/scikit-hep/blob/master/tests/conftest.py`, `https://github.com/scikit-hep/scikit-hep/blob/master/tests/visual/test_hists.py`, and related material. Hope that helps?

**Comment by eduardo-rodrigues** [comment]
Of course a toy manager would work, as you say. You could have your class as a fixture with 'session? scope ...

**Comment by marinang** [comment]
@eduardo-rodrigues the generations of toys (plus fit / plus scan) easily takes hour even for the simple example like this one with a (gaussian signal over an exponential background)[https://nbviewer.jupyter.org/github/scikit-hep/hepstats/blob/4ff3efe70f600b05e8db5f1f98f75f1da79178e6/notebooks/hypotests/upperlimit_freq_zfit.ipynb].

**Comment by eduardo-rodrigues** [comment]
OK, then you have to store the data somewhere. Fair enough. You can still use a little fixture to make the data available in the tests, though …

I realise that `scikit-hep-testdata` needs to be worked on to work properly, but it might be a use-case for you? Basically the package is meant to store big-ish data files for tests. @benkrikler wanted to get back to the package and your use-case seems like a good reason to … ;-)

**Comment by marinang** [comment]
I am not sure whether it is useful of not to put the results of pseudo-experiments for a particular statistical model  in `scikit-hep-testdata`.  I have checked what I was doing with [lauztat](https://github.com/marinang/lauztat/tree/v1.1.1/tests/hypotests), I was using hdf5 files and for an upper limit calculation the file size is a few MB.

**Comment by eduardo-rodrigues** [comment]
That's a good point. Fair enough.

**Comment by mayou36** [comment]
I expect this file to be in fact quite small, with about 100-1000 numerical entries, do you agree? This would also allow to e.g. use yaml?

And it seems related, just on a lower level, to #19. Does it make sense to mix the two

**Comment by marinang** [comment]
@mayou36 from experience to get a sensible result 1000 is the bare minimum. For each toy you need to store 4 numbers (more if you do a 2D confidence interval). For a 3 sigma evidence 1000 is order of toys you want, but imagine for a 5 sigma discovery.

For upper limits ~ 1000 toys is fine but you have to multiply this by the number of values you scan for the POI (usually >= 10).

So I don’t think it is that small. For that regard I am a bit hesitant whether to store that in the same yml file where the stat results are stored.

**Comment by mayou36** [comment]
Yes, true, that seems like too much, at least in human-readable format. yaml offers also the storage of arrays and there are some formats that combine yaml with hdf5 such as ASDF

**Comment by marinang** [comment]
There is still one issue, when the toys are stored there is no information stored about what model was used to generate the toys. So the user would have to be careful to match the toys with the corresponding loss in the `FrequentistCalculator`. 
So I just got and idea, given that you are working on likelihood storage, would it make sense to allow some extra item to be written without interfering with the core content of the file? 
This would allow to match likelihood to toys in one single file, then I could add a classmethod `FrequentistCalculator.from_yaml(cls, yaml_file, fitting_backend)`.

**Comment by mayou36** [comment]
That would be great of course! It is though only a thing wrapper around
`model = fitting_backend.from_yaml(...)`, right? So we would need to get that up first.

For the moment being, we may can leave the coordination to the user and let him name things properly?

**Comment by marinang** [comment]
Or better `loss = fitting_backend.from_yaml(...)`.

> For the moment being, we may can leave the coordination to the user and let him name things properly?

Yes for the moment we can only do that.

**Comment by eduardo-rodrigues** [comment]
Seems that at least for a first implementation the proposal to use hdf5 is viable. I would start an implementation with that.

The point you made in connection with likelihood storage and fitting tools in general is not to be undermined, see my comment in https://github.com/scikit-hep/hepstats/issues/19. :+1: on your proposal.
